### PR TITLE
Implementing vdn report using analytics data product framework

### PR DESCRIPTION
**Pull request by ishawakankar**

### CODE REVIEW

File: data-products/src/main/scala/org/sunbird/analytics/sourcing/SourcingMetrics.scala [hunk 1]
```diff
@@ -0,0 +1,173 @@
...
+  val sunbirdHierarchyStore: String = AppConf.getConfig("course.metrics.cassandra.sunbirdHierarchyStore")
+
+  // $COVERAGE-OFF$ Disabling scoverage for main method
+  def main(config: String)(implicit sc: Option[SparkContext], fc: Option[FrameworkContext]): Unit = {
+    JobLogger.log(s"Started execution - $jobName",None, Level.INFO)
+    implicit val sparkContext: SparkContext = getReportingSparkContext(JSONUtils.deserialize[JobConfig](config))
+    implicit val frameworkContext: FrameworkContext = getReportingFrameworkContext()
+    val readConsistencyLevel: String = AppConf.getConfig("course.metrics.cassandra.input.consistency")
+    val sparkConf = sparkContext.getConf
+      .set("es.write.operation", "upsert")
```
Comment (SanthoshVasabhaktula) [inline data-products/src/main/scala/org/sunbird/analytics/sourcing/SourcingMetrics.scala:40]: We don't write any logic to ES

File: data-products/src/main/scala/org/sunbird/analytics/sourcing/SourcingMetrics.scala [hunk 2]
```diff
@@ -0,0 +1,173 @@
...
+                             name: String, status: String, contentType: Option[String], leafNodesCount: Int, lastUpdatedOn: String,
+                             depth: Int, createdOn: String, children: Option[List[TextbookHierarchy]], index: Int, parent: String)
+case class FinalReport(identifier: String,l1identifier: String,board: String, medium: String, grade: String, subject: String, name: String, chapters: String, channel: String, totalChapters: String, slug:String)
+case class TextbookResponse(l1identifier:String,board: String, medium: String, grade: String, subject: String, name: String, chapters: String, channel: String)
+
+object SourcingMetrics extends optional.Application with IJob with BaseReportsJob {
+
+  implicit val className = "org.sunbird.analytics.job.SourcingMetrics"
+  val jobName: String = "Sourcing Metrics Job"
+  val sunbirdHierarchyStore: String = AppConf.getConfig("course.metrics.cassandra.sunbirdHierarchyStore")
```
Comment (SanthoshVasabhaktula) [inline data-products/src/main/scala/org/sunbird/analytics/sourcing/SourcingMetrics.scala:31]: How do we get the cassandra ip from? Configure a cluster and let the cluster ip be input from the jobConfig

File: data-products/src/main/scala/org/sunbird/analytics/sourcing/SourcingMetrics.scala [hunk 3]
```diff
@@ -0,0 +1,173 @@
...
+    JobLogger.log(s"Started execution - $jobName",None, Level.INFO)
+    implicit val sparkContext: SparkContext = getReportingSparkContext(JSONUtils.deserialize[JobConfig](config))
+    implicit val frameworkContext: FrameworkContext = getReportingFrameworkContext()
+    val readConsistencyLevel: String = AppConf.getConfig("course.metrics.cassandra.input.consistency")
+    val sparkConf = sparkContext.getConf
+      .set("es.write.operation", "upsert")
+      .set("spark.cassandra.input.consistency.level", readConsistencyLevel)
+    val spark = SparkSession.builder.config(sparkConf).getOrCreate()
+
+    generateSourcingMetrics(spark, config)
```
Comment (SanthoshVasabhaktula) [inline data-products/src/main/scala/org/sunbird/analytics/sourcing/SourcingMetrics.scala:44]: Please follow the same naming convention as exhaust jobs. main() -> execute() -> process() etc

File: data-products/src/main/scala/org/sunbird/analytics/sourcing/SourcingMetrics.scala [hunk 4]
```diff
@@ -0,0 +1,173 @@
...
+  }
+
+  // $COVERAGE-ON$ Enabling scoverage for all other functions
+  def generateSourcingMetrics(spark: SparkSession, config: String)(implicit sc: SparkContext,fc: FrameworkContext): Unit = {
```
Comment (SanthoshVasabhaktula) [inline data-products/src/main/scala/org/sunbird/analytics/sourcing/SourcingMetrics.scala:48]: Why do we need to pass in both spark and sparkcontext? Pass SparkSession, Config and FC as implicit variables

File: data-products/src/main/scala/org/sunbird/analytics/sourcing/SourcingMetrics.scala [hunk 5]
```diff
@@ -0,0 +1,173 @@
...
+    val conf = JSONUtils.deserialize[Map[String,AnyRef]](config)
+    val textbooks = TextBookUtils.getTextBooks(conf, RestUtil)
```
Comment (SanthoshVasabhaktula) [inline data-products/src/main/scala/org/sunbird/analytics/sourcing/SourcingMetrics.scala:50]: Why do we need to pass static classes as method parameters?

File: data-products/src/main/scala/org/sunbird/analytics/sourcing/SourcingMetrics.scala [hunk 6]
```diff
@@ -0,0 +1,173 @@
...
+
+    val textbookReportData = getTextbookInfo(textbooks, spark, config)
+    val textbookReports = textbookReportData._1.toDF()
+    val tenantInfo = getTenantInfo(RestUtil).toDF()
+    val report = textbookReports.join(tenantInfo,
+      textbookReports.col("channel") === tenantInfo.col("id"),"left")
+    generateSourcingReport(spark, textbookReportData._2.toDF(), report, config)
+  }
+
+  def generateSourcingReport(spark: SparkSession, contentdf: DataFrame, report: DataFrame, config: String): Unit = {
```
Comment (SanthoshVasabhaktula) [inline data-products/src/main/scala/org/sunbird/analytics/sourcing/SourcingMetrics.scala:62]: SparkSession, FrameworkContext and JobConfig should be passed as implicit parameters

Comment (SanthoshVasabhaktula) [inline data-products/src/main/scala/org/sunbird/analytics/sourcing/SourcingMetrics.scala:62]: generateReport is sufficient. why do we need to distinguish as sourcing report?

File: data-products/src/main/scala/org/sunbird/analytics/sourcing/SourcingMetrics.scala [hunk 7]
```diff
@@ -0,0 +1,173 @@
...
+    import spark.implicits._
+    val contentChapter = contentdf.groupBy("identifier","l1identifier")
+      .pivot(concat(lit("Number of "), col("contentType"))).agg(count("l1identifier"))
+    val contentTb = contentdf.groupBy("identifier")
+      .pivot(concat(lit("Number of "), col("contentType"))).agg(count("identifier"))
+    val configMap = JSONUtils.deserialize[Map[String,AnyRef]](config)
+    val reportConfig = configMap("modelParams").asInstanceOf[Map[String, AnyRef]]("reportConfig").asInstanceOf[Map[String, AnyRef]]
+    val reportPath = reportConfig.getOrElse("reportPath","sourcing").asInstanceOf[String]
+    val reportKey = reportConfig.getOrElse("reportKey","reports").asInstanceOf[String]
+    val storageConfig = getStorageConfig(reportKey, "")
```
Comment (SanthoshVasabhaktula) [inline data-products/src/main/scala/org/sunbird/analytics/sourcing/SourcingMetrics.scala:72]: take in container and secret keys from the job config as well

File: data-products/src/main/scala/org/sunbird/analytics/sourcing/SourcingMetrics.scala [hunk 8]
```diff
@@ -0,0 +1,173 @@
...
+    saveReportToBlob(textbookReport, JSONUtils.serialize(reportConfig), storageConfig, "TextbookLevel", reportPath)
+
+    val chapterReport = report.join(contentChapter, Seq("identifier","l1identifier"),"left")
+      .drop("identifier","l1identifier","channel","id","totalChapters")
+      .orderBy('medium,split(split('grade,",")(0)," ")(1).cast("int"),'subject,'name,'chapters)
+    JobLogger.log(s"$jobName: extracted chapter and textbook reports", None, INFO)
+    saveReportToBlob(chapterReport, JSONUtils.serialize(reportConfig), storageConfig, "ChapterLevel", reportPath)
+  }
+
+  def getTextbookInfo(textbooks: List[TextbookData], spark: SparkSession, config: String)(implicit sc: SparkContext,fc: FrameworkContext): (List[TextbookReportResult],List[ContentReportResult]) = {
```
Comment (SanthoshVasabhaktula) [inline data-products/src/main/scala/org/sunbird/analytics/sourcing/SourcingMetrics.scala:87]: Same as above. Pass in SparkSession, FrameworkContext and Config as implicit parameters

File: data-products/src/test/scala/org/sunbird/analytics/sourcing/TestSourcingMetrics.scala [hunk 9]
```diff
@@ -0,0 +1,84 @@
...
+    SourcingMetrics.generateSourcingMetrics(spark, config)
```
Comment (SanthoshVasabhaktula) [inline data-products/src/test/scala/org/sunbird/analytics/sourcing/TestSourcingMetrics.scala:60]: What are you asserting for here?

File: data-products/src/test/scala/org/sunbird/analytics/sourcing/TestSourcingMetrics.scala [hunk 10]
```diff
@@ -0,0 +1,84 @@
...
+    val hierarchy = JSONUtils.deserialize[TextbookHierarchy](data.hierarchy)
+    val report = SourcingMetrics.generateReport(List(hierarchy),List(), List(),hierarchy,List(),List("","0"))
+    val df = report._1.toDF().withColumn("slug",lit("test-slug")).withColumn("reportName",lit("TextbookReport"))
+    val configMap = JSONUtils.deserialize[Map[String,AnyRef]](config)
+    val reportConfig = configMap("modelParams").asInstanceOf[Map[String, AnyRef]]("reportConfig").asInstanceOf[Map[String, AnyRef]]
+    SourcingMetrics.saveReportToBlob(df,JSONUtils.serialize(reportConfig),StorageConfig("local","test","Textbook",Option(""),Option("")),"TextbookReport", "sourcing")
+
+    report._1.length should be (1)
+    report._2.length should be (2)
+    report._2.map(f => f.contentType should not be("TextbookUnit"))
```
Comment (SanthoshVasabhaktula) [inline data-products/src/test/scala/org/sunbird/analytics/sourcing/TestSourcingMetrics.scala:81]: Not enough assertions. Need to assert for each column and the variations of column data

**Review by SanthoshVasabhaktula — CHANGES_REQUESTED** [review]

File: data-products/src/main/scala/org/sunbird/analytics/sourcing/SourcingMetrics.scala (hunk 1 above)
Comment (ishawakankar) [reply data-products/src/main/scala/org/sunbird/analytics/sourcing/SourcingMetrics.scala]: Done

File: data-products/src/main/scala/org/sunbird/analytics/sourcing/SourcingMetrics.scala (hunk 3 above)
Comment (ishawakankar) [reply data-products/src/main/scala/org/sunbird/analytics/sourcing/SourcingMetrics.scala]: Done

File: data-products/src/main/scala/org/sunbird/analytics/sourcing/SourcingMetrics.scala (hunk 5 above)
Comment (ishawakankar) [reply data-products/src/main/scala/org/sunbird/analytics/sourcing/SourcingMetrics.scala]: Removed the logic and modified the dependent jobs

File: data-products/src/test/scala/org/sunbird/analytics/sourcing/TestSourcingMetrics.scala (hunk 9 above)
Comment (ishawakankar) [reply data-products/src/test/scala/org/sunbird/analytics/sourcing/TestSourcingMetrics.scala]: Added assertions

File: data-products/src/test/scala/org/sunbird/analytics/sourcing/TestSourcingMetrics.scala (hunk 10 above)
Comment (ishawakankar) [reply data-products/src/test/scala/org/sunbird/analytics/sourcing/TestSourcingMetrics.scala]: Added assertions for all columns

File: data-products/src/main/scala/org/sunbird/analytics/sourcing/SourcingMetrics.scala (hunk 2 above)
Comment (ishawakankar) [reply data-products/src/main/scala/org/sunbird/analytics/sourcing/SourcingMetrics.scala:31]: The cassandra ip is being fetched from jobConfig variable itself in getReportingSparkContext(jobConfig)

**Review by SanthoshVasabhaktula — APPROVED** [review]

**Pull request merged**